### ISSUE TITLE
To share my code "use tensorrt speed up mask-rcnn to 90fps"

### ISSUE THREAD

**Issue by chenxinfeng4** [open-mmlab/mmdetection#10158]
I've done some code to speed up detection model in video processing. I've improve the mask-rcnn from 14fps to 90fps in result. I want to share my code.

我改进了 mmdetection 在视频上 instance segmentation 推断的速度。在我的项目中可以达到mask rcnn 90fps，够用了。我认为这个改进对工业界比较重要，因而想把思路和代码分享出去。不知道你们怎么评价这个工作的重要性，以及是否值得分享？

**Describe the feature**
Speed up detection model in video processing using tensorrt & GPU asynchronization.

**Motivation**
I'm the author of #7832 , `mmdetection/demo/video_gpuaccel_demo.py`. I'm interested in speeding up instance segmentation in **video processing**.  Recently, I used tensorrt and some tricks which can speed up mask-rcnn to 90+fps (RTX3090, fp16, 816x512image), and take sufficient GPU occupancy (90% ). Other model may also be compatible. It would be useful for industry.

**Result** The video processing in RTX 3090. Mask RCNN.  832x512 imagesize.
<img width="715" alt="图片" src="https://user-images.githubusercontent.com/17546446/232238005-ed0ef999-5996-4cec-8070-7d1226383a7b.png">

**How to speed up**
1. Use tensorrt fp16. (Similar to mmdeploy)
2. Speed up video frame preprocess, because frames have the same size (mmdetection/demo/video_gpuaccel_demo.py)
3. Use GPU CPU asynchronization. Take the most use of GPU without waiting CPU.
4. Light weight `mmdet.models.roi_heads.mask_heads.FCNMaskHead.get_seg_masks`. Use cv2 instead of pytorch to restore the mask.

**Demo: core code**
3. GPU CPU asynchronization
```python
#the `model` is a tensorrt proxy.
with torch.cuda.device(self.cuda), torch.no_grad():
    for iframe in tqdm.trange(maxlen, position=self.id, desc=f"[{self.id}]"):
        ret, frame = vid.read()                    #0:n      | CPU
        outputs = model_output_detach(gpu_outputs) #-1:n-1   | GPU force synchronization
        data = process_img(frame, img_metas)       #0:n      | CPU
        gpu_outputs= model.forward_test(data['img']) #0:n    | GPU start asynchronization, GPU heavy
        result = model_output_result(outputs, img_metas, model.CLASSES) #-1:n-1  | CPU heavy
        if iframe>0:
            q.put((iframe-1, result))              #-1:n-1   | multiprocessing CPU postprocessing, CPU heavy
```

4. Light weight inference.
```python
def _do_paste_mask(masks, boxes, img_h, img_w, skip_empty=True):
    """Paste instance masks according to boxes.

    This implementation is modified from
    https://github.com/facebookresearch/detectron2/
   
   inference的时候可以用CV2来替换，与原版在“边界附近处理”略有差异
   """
    # batch size 必须是1
     masks = masks.numpy() # 1. 将torch的Tensor张量类型替换为对应的numpy数组。
    boxes = boxes.numpy()
    assert skip_empty
    x0_int, y0_int = np.maximum(
        np.floor(np.min(boxes[:,:2], axis=0)) - 1, 0).astype(np.int64)
    x1_int = np.minimum(np.ceil(np.max(boxes[:,2])) + 1, img_w).astype(np.int64)
    y1_int = np.minimum(np.ceil(np.max(boxes[:,3])) + 1, img_h).astype(np.int64)
    
    N, N2 = masks.shape[:2]
    assert N==1 and N2==1
    mask = np.squeeze(masks)
    mask_resized = cv2.resize(mask, (x1_int - x0_int, y1_int - y0_int))
    return mask_resized[None,...], (slice(y0_int, y1_int), slice(x0_int, x1_int))
```
